### ISSUE TITLE
Issues using error binding patterns as members of other binding patterns with var

### ISSUE THREAD

**Issue by MaryamZi** [ballerina-platform/ballerina-lang#19619]
**Description:**
$title.

**Steps to reproduce:**
```ballerina
public function main() {
    error e = error("reason", message = "msg");
    var [i, error(r, message = message)] = [1, e]; // invalid literal for type 'tuple binding pattern'
    // [int, error] [i, error(r, message = message)] = [1, e]; // but this works
}
```

**Affected Versions:**
1.0.1

**Comment by MaryamZi** [comment]
Getting a bad, sad error with the following

```ballerina
public function main() {
    string r;
    string? message;
    int i;

    error e = error("r2", message = "msg");
    [i, error(r, message = message)] = [1, e];
}
```

ballerina-internal.log
```
[2019-10-28 17:43:43,277] SEVERE {b7a.log.crash} - $type$ 
java.lang.NoSuchFieldError: $type$
	at xyz.main(xyz.bal:7)
	at ___init.$lambda$main$(.)
	at org.ballerinalang.jvm.scheduling.SchedulerItem.execute(Scheduler.java:408)
	at org.ballerinalang.jvm.scheduling.Scheduler.run(Scheduler.java:202)
	at org.ballerinalang.jvm.scheduling.Scheduler.runSafely(Scheduler.java:175)
	at java.lang.Thread.run(Thread.java:748)
 
[2019-10-28 17:43:43,282] SEVERE {b7a.log.crash} - $type$ 
java.lang.NoSuchFieldError: $type$
	at xyz.main(xyz.bal:7)
	at ___init.$lambda$main$(.)
	at org.ballerinalang.jvm.scheduling.SchedulerItem.execute(Scheduler.java:408)
	at org.ballerinalang.jvm.scheduling.Scheduler.run(Scheduler.java:202)
	at org.ballerinalang.jvm.scheduling.Scheduler.runSafely(Scheduler.java:175)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by MaryamZi** [comment]
There's a compilation error when trying to use an error b-p with destructuring assignment with the error constructor.

But, when the same is part of a list b-p results in a bad, sad error.

```ballerina
public function main() {
    string r;
    string? message;
    int i;
    
    // error(r, message = message) = error("r2", message = "hi"); // error constructor expression is not supported for error binding pattern

    [i, error(r, message = message)] = [1, error("r2", message = "hi")]; // when the error b-p is part of a list b-p
}
```
ballerina-internal.log
```cmd
[2019-10-28 19:38:21,919] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.isExternalLangLibFunction(TaintAnalyzer.java:1138)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:1072)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangInvocation.accept(BLangInvocation.java:119)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:984)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangListConstructorExpr.accept(BLangListConstructorExpr.java:51)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:835)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangTupleDestructure.accept(BLangTupleDestructure.java:66)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeNode(TaintAnalyzer.java:1695)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:552)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangBlockStmt.accept(BLangBlockStmt.java:56)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeNode(TaintAnalyzer.java:1695)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeReturnTaintedStatus(TaintAnalyzer.java:2124)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeReturnTaintedStatus(TaintAnalyzer.java:2046)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeAllParamsUntaintedReturnTaintedStatus(TaintAnalyzer.java:1982)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visitInvokable(TaintAnalyzer.java:1902)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:401)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.lambda$analyzeNode$6(TaintAnalyzer.java:335)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeNode(TaintAnalyzer.java:332)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyze(TaintAnalyzer.java:299)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:283)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:161)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyze(TaintAnalyzer.java:265)
	at org.wso2.ballerinalang.compiler.CompilerDriver.taintAnalyze(CompilerDriver.java:283)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compile(CompilerDriver.java:244)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackageSymbol(CompilerDriver.java:216)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackage(CompilerDriver.java:131)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:173)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:178)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:93)
	at org.wso2.ballerinalang.compiler.Compiler.build(Compiler.java:101)
	at org.ballerinalang.packerina.task.CompileTask.execute(CompileTask.java:51)
	at org.ballerinalang.packerina.TaskExecutor.executeTasks(TaskExecutor.java:38)
	at org.ballerinalang.packerina.cmd.RunCommand.execute(RunCommand.java:268)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.tool.Main.main(Main.java:59)
```

**Comment by rdhananjaya** [comment]
2nd and 3rd comments are fixed in #19660 
Original (1st) issue is a validation in list constructors/destructure, this is probably due to some ambiguity in deciding between array vs tuple.

I think @grainier can confirm this.

**Comment by grainier** [comment]
> Original (1st) issue is a validation in list constructors/destructure, this is probably due to some ambiguity in deciding between array vs tuple.

Yes. This is correct. If it's declared with var and RHS is list constructor expr, it is considered ambiguous. 

Further, refer to https://github.com/ballerina-platform/ballerina-spec/issues/392 on resolving contextually expected type on tuples/arrays.
